### DISCUSSION TITLE
Any nice way to inspect build configuration from binaries?

### DISCUSSION THREAD

**Discussion by EliahKagan** [GitoxideLabs/gitoxide#1492] [Q&A]
Git has `git version --build-options`, which gives some information about the build, though not which libraries it is linked against:

```text
ek@Glub:~$ git version --build-options
git version 2.34.1
cpu: x86_64
no commit associated with this build
sizeof-long: 8
sizeof-size_t: 8
shell-path: /bin/sh
```

Libraries Git uses that are shared libraries can be inspected with `ldd`:

```text
ek@Glub:~$ ldd "$(command -v git)"
        linux-vdso.so.1 (0x00007ffe61462000)
        libpcre2-8.so.0 => /lib/x86_64-linux-gnu/libpcre2-8.so.0 (0x00007fe87a15f000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007fe87a143000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fe879f1a000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fe87a5ad000)
```

(I don't know if `git` provides any convenient way to list which features that require library support are available when the library is statically linked.)

**Does `gix` or `ein` have any way to obtain this kind of information about itself?** It would be useful to know details corresponding to the output of `git version --build-options` where applicable, but even more so to know which [feature flags](https://docs.rs/crate/gitoxide/latest/features) `gix` and `ein` are compiled with.

Even if `gitoxide` does not have corresponding functionality, is there an easy way to check this by running or inspecting the binaries? (I've opened this as a question rather than a feature request because I'm not sure I wish to request a change. Maybe this information is already readily available.)

**Comment by NobodyXu** [comment]
There's [verge ](https://docs.rs/vergen) which enables encoding the build environment and git sha in the binary.

cargo-binstall print these info on `-vV`.

There's also `cargo-auditable`, which encodes the dependencies in the binary for auditing, we were able to use it with cargo-zigbuild by setting `export CARGO=cargo-zigbuild`

**Reply by EliahKagan** [reply]
Thanks!

I had noticed at https://github.com/rust-secure-code/cargo-auditable/issues/124#issuecomment-2271580700 that, at some point, `cargo-auditable` had started working again for `gitoxide` builds, and I have verified locally that it works. That's something it seems to me we should use for the binaries produced by the release workflow, if possible, so that `cargo audit bin` can be used on them. I expect the value of being able to do that to increase over time as `gitoxide` is used more widely.

(For example, #1238 can't be a [container action](https://docs.github.com/en/actions/creating-actions/about-custom-actions#types-of-actions), because *(a)* it is to be a drop-in replacement for [actions/checkout](https://github.com/actions/checkout), so it has to be able to run on non-Linux runners, which [container actions can't](https://github.com/Byron/gitoxide/issues/1472#issuecomment-2254244666), and *(b)* a container action would have too much overhead to allow it to be at least as fast as actions/checkout for the common case of a shallow clone of a modestly sized repository. So it would have to use a separate executable, possibly even a build of the `gix` binary. But unlike `git`, most runners, including all current GitHub-hosted runners, don't have `gix` already. So it would often need to be downloaded. To speed that up, it would sometimes be cached, so it would benefit from being auditable.)

Nonetheless, I had not thought of how that would also, in effect, address much or all of what I'm asking for in this question. As for `verge`, I was completely unaware of it! I will look into it.

Currently all Linux targets in `release.yml` are built using `cross`, and I think that will probably remain the case for some of them (most, assuming more are added, which I intend to do, in the short term to add all remaining targets requested in #1242, and eventually I think perhaps a number of others, so #1238 can seamlessly support self-hosted runners on less common architectures such as s390x and riscv64). I'm not sure how easy or hard it is to use `cargo-auditable` together with `cross`. But if you have any thoughts about that, then I would definitely be interested.

**Reply by EliahKagan** [reply]
Searching further, I can see this is something you have already researched, and found solutions to, in https://github.com/rust-secure-code/cargo-auditable/issues/95 and https://github.com/cross-rs/cross/issues/1172.

Although not the only solution, it seems from https://github.com/cross-rs/cross/issues/1172#issuecomment-1546913099 that this is another area where `cargo-zigbuild` may carry advantages. I wonder if this is something that should be considered for use here (by `release.yml`).

**Reply by Byron** [reply]
I also think something like it would be very useful to have, as currently it only bakes in the git commit hash. For a BoM I'd also hope that the tools mentioned here could be made to work cross-platform. Lastly, feature-flags might be harder to do unless these options are passed via environment variables which could be observable during build time. 

Cargo [already sets a bunch of variables](https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-crates) at build time, but it doesn't look like top-level feature flags are among them.

Maybe having a BoM already is what's needed though, as feature-toggles usually also affect the crates that are compiled in.

**Reply by EliahKagan** [reply]
In the release workflow, since b84162d (#1486), we do use an environment variable for the feature flags:

https://github.com/Byron/gitoxide/blob/29898e3010bd3332418c683f2ac96aff5c8e72fa/.github/workflows/release.yml#L137

That is expanded in various places, including:

https://github.com/Byron/gitoxide/blob/29898e3010bd3332418c683f2ac96aff5c8e72fa/.github/workflows/release.yml#L174-L176

Although in current use it is always just expanded by the shell, it is an environment variable and not just a shell variable (i.e., the shell inherits it and thus automatically exports it), so I think subprocesses should see it, unless something is getting in the way of that. So maybe nothing has to be done for that specific requirement.

**Reply by Skgland** [reply]
> Cargo [already sets a bunch of variables](https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-crates) at build time, but it doesn't look like top-level feature flags are among them.

To at least get some feature information you could have something like a static storing whether the feature is enabled i.e.

```rust
static HAS_FEATURE_BAR: bool = cfg!(feature="bar");
```

**Reply by Byron** [reply]
Thanks for clarifying this. So it does seem that GitHub releases could definitely incorporate this knowledge, with `env_opt!` maybe, I'd definitely appreciate it, and it seems like a simple and portable first step. BoM could be a second one but is riskier due to an additional tool being involved.

**Reply by Byron** [reply]
> > Cargo [already sets a bunch of variables](https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-crates) at build time, but it doesn't look like top-level feature flags are among them.
> 
> To at least get some feature information you could have something like a static storing whether the feature is enabled i.e.
> 
> ```rust
> static HAS_FEATURE_BAR: bool = cfg!(feature="bar");
> ```

I thought about this but rejected it for myself as it needs maintenance, and I know I will forget it, for sure 😅. Using specific environment variables that are known to exist in GitHub releases seems like the way to go.

**Reply by Skgland** [reply]
> Cargo [already sets a bunch of variables](https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-crates) at build time, but it doesn't look like top-level feature flags are among them.

For [build script](https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-build-scripts) cargo sets `CARGO_FEATURE_<name>` for each activated feature, so a build script could collect all available environment variables starting with `CARGO_FEATURE_` to get a list of enabled features, but given that you currently don't appear to be using build scripts this is probably not useful. 

> CARGO_FEATURE_\<name> — For each activated feature of the package being built, this environment variable will be present where \<name> is the name of the feature uppercased and having - translated to _.

**Reply by NobodyXu** [reply]
> but given that you currently don't appear to be using build scripts this is probably not useful.

I think build-script is required if we want gitoxide to work on 32-bit windows cargo-bins/cargo-quickinstall#254

Windows requires you to embed a manifest in the binary for it to work.

Cargo-binstall uses build-script to generate a windows manifest and record build environment.

https://github.com/cargo-bins/cargo-binstall/blob/main/crates/bin/build.rs

**Reply by Byron** [reply]
There is indeed [a build script](https://github.com/Byron/gitoxide/blob/2189cee47f99350b368390eaa2a01961bb77c250/build.rs#L1-L0) for the binaries already, which could then aggregate cargo features and make them available in another environment variable, thanks @Skgland for the suggestion.

This makes me think a general embedding of features to present when running `gix -v` or `gix -vv` would be quite easy to accomplish.

**Reply by EliahKagan** [reply]
In https://github.com/GitoxideLabs/gitoxide/discussions/1492#discussioncomment-10270715 I had made this strange claim:

> I had noticed at https://github.com/rust-secure-code/cargo-auditable/issues/124#issuecomment-2271580700 that, at some point, `cargo-auditable` had started working again for `gitoxide` builds, and I have verified locally that it works.

I wish I remembered what, exactly, I had tested. It may be that I tested a patched `gitoxide` and/or a patched `cargo-auditable`, or that I tested with `gitoxide` features that were able to build even when others were not, or maybe I only tested on one platform. In any case, as noted in https://github.com/rust-secure-code/cargo-auditable/issues/124#issuecomment-2271646061 and https://github.com/rust-secure-code/cargo-auditable/issues/124#issuecomment-2271664929, it wasn't really fixed at that time.

However, since #1929, it really does work (https://github.com/rust-secure-code/cargo-auditable/issues/124#issuecomment-2781505478). Running `cargo auditable build -r` will build `gitoxide` with no errors.

Unfortunately, `cargo-auditable` does not readily work with `cross`, and modifying the Docker image used by `cross` is not sufficient to fix this, at least not by doing it in the ways that had, at one time, been expected to work. (Per https://github.com/rust-secure-code/cargo-auditable/issues/95#issuecomment-2781520669, the approaches mentioned in https://github.com/GitoxideLabs/gitoxide/discussions/1492#discussioncomment-10271039 seem to be untested and infeasible as described.)

@NobodyXu If I understand correctly what you were saying in https://github.com/GitoxideLabs/gitoxide/discussions/1492#discussioncomment-10270462, `cargo auditable` works--or had at some point worked--when using `cargo-zigbuild`, which could be used instead of `cross`. But https://github.com/rust-secure-code/cargo-auditable/issues/179 and https://github.com/rust-secure-code/cargo-auditable/issues/95#issuecomment-2781528677 indicate that this may not be working. Is this something that you are currently able to use?

Recently published packages from QuickInstall don't seem to have `cargo-auditable` metadata, but I haven't examined this comprehensively and I also don't know if that's the setting where you had used this.

**Reply by NobodyXu** [reply]
> @NobodyXu If I understand correctly what you were saying in https://github.com/GitoxideLabs/gitoxide/discussions/1492#discussioncomment-10270462, cargo auditable works--or had at some point worked--when using cargo-zigbuild, which could be used instead of cross. Is this something that you are currently able to use?

yes, cargo-binstall is still using it https://github.com/cargo-bins/cargo-binstall/blob/63aaa5c1932cebabc34eceda9d92a70215dcead6/justfile#L20

> Recently published packages from QuickInstall don't seem to have cargo-auditable metadata, but I haven't examined this comprehensively and I also don't know if that's the setting where you had used this.

I removed cargo-auditable for quickinstall due to some quirk/bugs
